File: backend/api/workspace.py
"""工作区 API."""

from fastapi import APIRouter
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime

router = APIRouter()


class PendingBook(BaseModel):
    """待下载书籍."""
    id: int
    title: str
    author: str
    language: str = "en"
    selected: bool = True


class DownloadResult(BaseModel):
    """下载结果."""
    bookId: int
    title: str
    success: bool
    filePath: Optional[str] = None
    error: Optional[str] = None


class Batch(BaseModel):
    """下载批次."""
    id: int
    name: str
    createdAt: str
    status: str
    total: int
    success: int
    failed: int
    results: List[DownloadResult] = []


class WorkspaceData(BaseModel):
    """工作区数据."""
    version: str = "1.0"
    createdAt: str
    updatedAt: str
    pendingDownloads: List[PendingBook] = []
    currentBatch: Optional[int] = None
    batches: List[Batch] = []


class WorkspaceStatus(BaseModel):
    """工作区状态."""
    isOpen: bool
    path: Optional[str] = None
    data: Optional[WorkspaceData] = None


# 临时存储 (实际应该存储在文件系统)
_current_workspace: dict = {}


@router.post("/open")
async def open_workspace(path: str):
    """
    打开工作区.

    初始化或加载工作区数据。
    """
    global _current_workspace

    # TODO: 实际从文件系统读取
    now = datetime.now().isoformat()

    _current_workspace = {
        "path": path,
        "data": WorkspaceData(
            createdAt=now,
            updatedAt=now
        ).model_dump()
    }

    return _current_workspace["data"]


@router.get("/status", response_model=WorkspaceStatus)
async def get_status():
    """获取工作区状态."""
    if not _current_workspace:
        return WorkspaceStatus(isOpen=False)

    return WorkspaceStatus(
        isOpen=True,
        path=_current_workspace.get("path"),
        data=WorkspaceData(**_current_workspace.get("data", {}))
    )


@router.post("/save")
async def save_workspace(data: WorkspaceData):
    """保存工作区数据."""
    global _current_workspace

    if not _current_workspace:
        return {"success": False, "error": "没有打开的工作区"}

    data.updatedAt = datetime.now().isoformat()
    _current_workspace["data"] = data.model_dump()

    # TODO: 实际保存到文件系统

    return {"success": True}

File: electron/workspace.ts
import fs from 'fs'
import path from 'path'

const WORKSPACE_DIR = '.bookweaver'
const CONFIG_FILE = 'config.json'
const WORKSPACE_FILE = 'workspace.json'
const AI_CONTEXT_FILE = 'ai_context.json'
const DOWNLOADS_DIR = 'downloads'

interface WorkspaceData {
  version: string
  createdAt: string
  updatedAt: string
  pendingDownloads: PendingBook[]
  currentBatch: number | null
  batches: Batch[]
}

interface PendingBook {
  id: number
  title: string
  author: string
  language: string
  selected: boolean
}

interface Batch {
  id: number
  name: string
  createdAt: string
  status: 'downloading' | 'completed' | 'failed'
  total: number
  success: number
  failed: number
  results: DownloadResult[]
}

interface DownloadResult {
  bookId: number
  title: string
  success: boolean
  filePath?: string
  error?: string
}

interface AIContext {
  history: Array<{ role: string; content: string }>
  bookList: PendingBook[]
}

interface Config {
  llm: {
    apiKey: string
    model: string
    baseUrl: string
    temperature: number
    maxTokens: number
  }
  download: {
    concurrent: number
    timeout: number
  }
}

const DEFAULT_CONFIG: Config = {
  llm: {
    apiKey: '',
    model: 'qwen3.5-plus',
    baseUrl: 'https://dashscope.aliyuncs.com/compatible-mode/v1',
    temperature: 0.7,
    maxTokens: 2000
  },
  download: {
    concurrent: 3,
    timeout: 30
  }
}

const DEFAULT_WORKSPACE: WorkspaceData = {
  version: '1.0',
  createdAt: new Date().toISOString(),
  updatedAt: new Date().toISOString(),
  pendingDownloads: [],
  currentBatch: null,
  batches: []
}

const DEFAULT_AI_CONTEXT: AIContext = {
  history: [],
  bookList: []
}

export class WorkspaceManager {
  private basePath: string
  private workspacePath: string

  constructor(folderPath: string) {
    this.basePath = folderPath
    this.workspacePath = path.join(folderPath, WORKSPACE_DIR)
  }

  async initialize(): Promise<void> {
    // 创建 .bookweaver 目录
    if (!fs.existsSync(this.workspacePath)) {
      fs.mkdirSync(this.workspacePath, { recursive: true })
    }

    // 创建 downloads 子目录
    const downloadsPath = path.join(this.workspacePath, DOWNLOADS_DIR)
    if (!fs.existsSync(downloadsPath)) {
      fs.mkdirSync(downloadsPath, { recursive: true })
    }

    // 初始化默认文件
    const configPath = path.join(this.workspacePath, CONFIG_FILE)
    if (!fs.existsSync(configPath)) {
      this.saveConfig(DEFAULT_CONFIG)
    }

    const workspaceFilePath = path.join(this.workspacePath, WORKSPACE_FILE)
    if (!fs.existsSync(workspaceFilePath)) {
      this.save(DEFAULT_WORKSPACE)
    }

    const aiContextPath = path.join(this.workspacePath, AI_CONTEXT_FILE)
    if (!fs.existsSync(aiContextPath)) {
      this.saveAIContext(DEFAULT_AI_CONTEXT)
    }
  }

  getData(): WorkspaceData {
    const filePath = path.join(this.workspacePath, WORKSPACE_FILE)
    try {
      const content = fs.readFileSync(filePath, 'utf-8')
      return JSON.parse(content)
    } catch {
      return DEFAULT_WORKSPACE
    }
  }

  save(data: unknown): void {
    const filePath = path.join(this.workspacePath, WORKSPACE_FILE)
    const workspaceData = {
      ...(data as WorkspaceData),
      updatedAt: new Date().toISOString()
    }
    fs.writeFileSync(filePath, JSON.stringify(workspaceData, null, 2))
  }

  getConfig(): Config {
    const filePath = path.join(this.workspacePath, CONFIG_FILE)
    try {
      const content = fs.readFileSync(filePath, 'utf-8')
      return { ...DEFAULT_CONFIG, ...JSON.parse(content) }
    } catch {
      return DEFAULT_CONFIG
    }
  }

  saveConfig(config: unknown): void {
    const filePath = path.join(this.workspacePath, CONFIG_FILE)
    fs.writeFileSync(filePath, JSON.stringify(config, null, 2))
  }

  getAIContext(): AIContext {
    const filePath = path.join(this.workspacePath, AI_CONTEXT_FILE)
    try {
      const content = fs.readFileSync(filePath, 'utf-8')
      return JSON.parse(content)
    } catch {
      return DEFAULT_AI_CONTEXT
    }
  }

  saveAIContext(context: unknown): void {
    const filePath = path.join(this.workspacePath, AI_CONTEXT_FILE)
    fs.writeFileSync(filePath, JSON.stringify(context, null, 2))
  }

  saveBatch(id: number, data: Batch): void {
    const downloadsPath = path.join(this.workspacePath, DOWNLOADS_DIR)
    const filePath = path.join(downloadsPath, `batch_${id}.json`)
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2))
  }

  getBatch(id: number): Batch | null {
    const downloadsPath = path.join(this.workspacePath, DOWNLOADS_DIR)
    const filePath = path.join(downloadsPath, `batch_${id}.json`)
    try {
      const content = fs.readFileSync(filePath, 'utf-8')
      return JSON.parse(content)
    } catch {
      return null
    }
  }

  getBasePath(): string {
    return this.basePath
  }

  getWorkspacePath(): string {
    return this.workspacePath
  }
}